File: src/main/java/su/sonoma/lostriver/block/ModPortalBlock.java
package su.sonoma.lostriver.block;


import net.minecraft.client.gui.font.providers.UnihexProvider;
import net.minecraft.core.BlockPos;
import net.minecraft.resources.ResourceKey;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.EntityDimensions;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.phys.BlockHitResult;
import su.sonoma.lostriver.dimension.B4546;
import su.sonoma.lostriver.portal.ModTeleporter;

public class ModPortalBlock extends Block {
    public ModPortalBlock(Properties pProperties) {
        super(pProperties);
    }

    @Override
    public InteractionResult use(BlockState pState, Level pLevel, BlockPos pPos, Player pPlayer, InteractionHand pHand, BlockHitResult pHit) {
        if (pPlayer.canChangeDimensions()) {
            handleKaupenPortal(pPlayer, pPos, pPlayer);
            return InteractionResult.SUCCESS;
        } else {
            return InteractionResult.CONSUME;
        }
    }

    private void handleKaupenPortal(Entity player, BlockPos pPos, Player player1) {
        if (player.level() instanceof ServerLevel serverlevel) {
            MinecraftServer minecraftserver = serverlevel.getServer();
            ResourceKey<Level> resourcekey = player.level().dimension() == B4546.INSTANCE.getB4546_LEVEL_KEY() ?
                    Level.OVERWORLD : B4546.INSTANCE.getB4546_LEVEL_KEY();

            ServerLevel portalDimension = minecraftserver.getLevel(resourcekey);
            if (portalDimension != null && !player.isPassenger()) {
                if(resourcekey == B4546.INSTANCE.getB4546_LEVEL_KEY()) {
                    player.changeDimension(portalDimension, new ModTeleporter(pPos, true));
                } else {
                    player.changeDimension(portalDimension, new ModTeleporter(pPos, false));
                }
            }
        }
    }
}

File: src/main/kotlin/su/sonoma/lostriver/block/custom/ModPortalBlock.kt
package su.sonoma.lostriver.block.custom

import net.minecraft.core.BlockPos
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.InteractionHand
import net.minecraft.world.InteractionResult
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.phys.BlockHitResult
import su.sonoma.lostriver.dimension.B4546
import su.sonoma.lostriver.portal.ModTeleporter


class ModPortalBlock(pProperties: Properties) : Block(pProperties) {
    override fun use(
        pState: BlockState,
        pLevel: Level,
        pPos: BlockPos,
        pPlayer: Player,
        pHand: InteractionHand,
        pHit: BlockHitResult,
    ): InteractionResult {
        if (pPlayer.canChangeDimensions()) {
            handlePortal(pPlayer, pPos)
            return InteractionResult.SUCCESS
        } else {
            return InteractionResult.CONSUME
        }
    }

    private fun handlePortal(player: Entity, pPos: BlockPos) {
        if (player.level() is ServerLevel) {
            val serverLevel = player.level() as ServerLevel
            val minecraftServer = serverLevel.server
            val resourceKey = if (player.level().dimension() == B4546.B4546_LEVEL_KEY)
                Level.OVERWORLD
            else
                B4546.B4546_LEVEL_KEY

            val portalDimension = minecraftServer.getLevel(resourceKey)
            if (portalDimension != null && !player.isPassenger) {
                player.changeDimension(portalDimension, ModTeleporter(pPos, resourceKey == B4546.B4546_LEVEL_KEY))
            }
        }
    }


}